STM32WB0: update debugging to provide instructions / work with SDK OpenOCD

mathieuchopstm opened the pull request:
OpenOCD support snippets (`<board>/support/openocd.cfg`) for Nucleo-WB05KZ and Nucleo-WB07CC were written with the expectation that ST OpenOCD was being used, but upstream and the Zephyr fork have everything needed to support STM32WB05 and STM32WB06/07 SoCs (a.k.a. BlueNRG-LPS and BlueNRG-LP respectively). Modify these boards' snippets so that they work out of the box with the SDK.

Support for STM32WB09 (a.k.a. BlueNRG-LPF) is only available on the [STMicroelectronics OpenOCD fork], with binaries shipped as part of STM32CubeIDE. Add warning about this and instructions on how to use CubeIDE's binaries to debug applications in the Nucleo-WB09KE board's documentation.

New support snippets tested OK on Linux with Zephyr SDK 0.17.0 and Windows using [official OpenOCD 0.12.0 release from GitHub](https://github.com/openocd-org/openocd/releases/tag/v0.12.0).

mathieuchopstm committed:
boards: st: nucleo_wb09ke: replace OpenOCD with pyOCD for debugging

pyOCD is installed along with Zephyr, and can debug apps on STM32WB09 once
the appropriate pack (already publicly available) is installed.

On the other hand, OpenOCD comes with the Zephyr SDK but only on Linux, and
both Zephyr's fork and upstream lack support for STM32WB09, which was added
by commit 89b053b660bcef803e19c89584de933158c59299 in ST's OpenOCD fork[1].
While pre-builts for this OpenOCD fork are provided by ST as part of other
software packages (e.g., STM32CubeIDE), it is less convenient to use and
the exact support script to use varies depending on which tool OpenOCD is
borrowed from.

Remove OpenOCD support for STM32WB09 entirely, replacing it with pyOCD, and
add instructions about pack installation in the board documentation.

While this also applies to flashing, STM32CubeProgrammer is the default
runner for this, and already supports STM32WB09 without any shenanigans
(i.e., a bare "west flash" will work with CubeProg) - as such, place the
notice only in the debugging section, which is the only operation CubeProg
does not support.

[1] https://github.com/STMicroelectronics/OpenOCD

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
boards: st: nucleo_wb07cc: make OpenOCD script compatible with Zephyr

The old OpenOCD script was expecting usage of ST's OpenOCD fork, but
the WB07 is already supported as "bluenrg-lp" in upstream OpenOCD and
the Zephyr fork provided in SDK.

Update the script to work properly with these two versions, instead of
requiring usage of ST's OpenOCD fork and scripts to function properly.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
boards: st: nucleo_wb05kz: make OpenOCD script compatible with Zephyr

The old OpenOCD script was expecting usage of ST's OpenOCD fork, but
the WB05 is already supported as "bluenrg-lps" in upstream OpenOCD and
the Zephyr fork provided in SDK.

Update the script to work properly with these two versions, instead of
requiring usage of ST's OpenOCD fork and scripts to function properly.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented:
> On Windows, you may want to test with https://gnutoolchains.com/arm-eabi/openocd/ instead, since it provides more recent & frequent builds compared to the official OpenOCD. Also it's what recommended in Zephyr docs https://docs.zephyrproject.org/latest/develop/flash_debug/host-tools.html#openocd-debug-host-tools

I didn't see this section in Zephyr's instructions so I wasn't aware this build was recommended... Despite not seeing a point in testing with this OpenOCD distribution, I did it anyways and can confirm that STM32WB09 doesn't work with it either. I haven't bothered testing WB05 or WB07 - they are [supported](https://github.com/openocd-org/openocd/commit/6bc0a77a6e1a1146c44785812595250857fc7307) [upstream](https://github.com/openocd-org/openocd/commit/ad5ca263e931792483cdb2a820b7aba3980829eb) since ages ago (note that both of these commits are in Zephyr's fork).

HoZHel commented on boards/st/nucleo_wb09ke/doc/index.rst:
```diff
@@ -119,6 +119,20 @@ You can debug an application in the usual way.  Here is an example for the
+
+      $ west debug --openocd="${STM32CubeIDE}/plugins/com.st.stm32cube.ide.mcu.externaltools.openocd.PLAT_X.X.XXX.XXXXXXXXXXXX/tools/bin/openocd"
+                   --openocd-search="${STM32CubeIDE}/plugins/com.st.stm32cube.ide.mcu.debug.openocd_X.X.XXX.XXXXXXXXXXXX/resources/openocd/st_scripts"
+
```
It's ok with me; however, an alternative option is to use the OpenOCD available in the [RF-Flasher utility](https://www.st.com/en/embedded-software/stsw-bnrgflasher.html). IMO, the instruction to use this one is easier.

mathieuchopstm replied:
Since RF-Flasher is a Windows-only utility, I would rather link to STM32CubeIDE which is cross-platform. Also, it seems RF-Flasher provides `bluenrg-x.cfg`, whereas CubeIDE provides only `stm32wb0x.cfg`, so a different `support.cfg` would be needed.

Your remark makes think about a solution I didn't think about thought; maybe only the ST OpenOCD *binary* is required, rather than the binary and scripts. On first look, it appears to work (which makes sense as I don't remember `bluenrg-x.cfg` being updated in the _add BlueNRG-LPF support_ commit), but I will do an additional round of testing to confirm this.

If it appears to work fine, I will update Nucleo-WB09KE's `support.cfg` and instructions to reflect this (i.e., only `west debug --openocd=XXX` rather than `west debug --openocd=XXX --openocd-search=XXX`), and I can add instructions for using RF-Flash Utility's OpenOCD as well.

erwango replied:
Can't we point to ST-OpenOCD version ?
I can see a commit adding "BlueNRG" support here: https://github.com/STMicroelectronics/OpenOCD/commit/89b053b660bcef803e19c89584de933158c59299

mathieuchopstm committed:
boards: st: nucleo_wb07cc: add pyOCD support

Add pyOCD as supported runner for the Nucleo-WB05KZ board, and a note about
pack installation to the board documentation.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
boards: st: nucleo_wb05kz: add pyOCD support

Add pyOCD as supported runner for the Nucleo-WB05KZ board, and a note about
pack installation to the board documentation.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented:
Following discussion w/ @erwango : since pyOCD supports STM32WB0 series (incl. STM32WB09) and comes installed in the Zephyr venv, replace OpenOCD with pyOCD for Nucleo-WB09KE, as it doesn't require installing third-party software, and solves the `support.cfg` question.

Also add pyOCD support to the other two WB0 boards.
